src/sympc/tensor/share_control.py

```diff
@@ -0,0 +1,249 @@
+            else:
+                share = secret - random_shares[i - 1]
+
+            shares.append(share)
+        return shares
+
+    def reconstruct(self, decode=True):
+        client_type = self.session.parties[0].class_name
+
+        def _request_and_get(share_ptr):
+            # If not VirtualMachine, we are in Duet and we need to request
+            # the data
+            share_ptr.request(block=True)
+            return share_ptr.get()
+
+        def _get(share_ptr):
+            return share_ptr.get()
+
+        request = None
+        if "VirtualMachine" in client_type:
+            request = _get
+        else:
+            request = _request_and_get
+
+        nr_parties = len(self.session.parties)
+        with ThreadPoolExecutor(
+            max_workers=nr_parties, thread_name_prefix="ast_request_and_get"
+        ) as executor:
+            futures = [executor.submit(request, share) for share in self.share_ptr]
+
+        local_shares = [f.result() for f in futures]
+
+        tensor_type = self.session.tensor_type
+        plaintext = sum(local_shares)
+
+        if decode:
+            fp_encoder = FixedPointEncoder(
+                base=self.session.config.encoder_base,
+                precision=self.session.config.encoder_precision,
+            )
+
+            plaintext = fp_encoder.decode(plaintext.tensor)
+        else:
+            plaintext = plaintext.tensor
+
+        return plaintext
+
+    def add(self, y):
+        return self.__apply_op(y, "add")
+
+    def sub(self, y):
+        return self.__apply_op(y, "sub")
+
+    def mul(self, y):
+        return self.__apply_op(y, "mul")
+
+    def div(self, y):
+        return self.__apply_op(y, "div")
+
+    def __apply_private_op(self, y, op_str):
+        if y.session.uuid != self.session.uuid:
+            raise ValueError(
+                f"Need same session {self.session.uuid} and {y.session.uuid}"
+            )
+
+        if op_str in {"mul"}:
+            from ..protocol import spdz
+
+            shares = spdz.mul_master(self, y, op_str)
+            result = ShareTensorCC(shares=shares, session=self.session)
+        elif op_str in {"sub", "add"}:
+            op = getattr(operator, op_str)
+            result = ShareTensorCC(session=self.session)
+            result.share_ptr = [
+                op(*share_tuple) for share_tuple in zip(self.share_ptr, y.share_ptr)
+            ]
+
+        return result
+
+    def __apply_public_op(self, y, op_str):
+        op = getattr(operator, op_str)
+        if op_str in {"mul"}:
+            shares = [op(share, y) for share in self.share_ptr]
+        else:
+            operands_shares = [y] + [0 for _ in range(len(self.share_ptr) - 1)]
+            shares = [
+                op(*tuple_shares)
+                for tuple_shares in zip(self.share_ptr, operands_shares)
+            ]
+
+        result = ShareTensorCC(shares=shares, session=self.session)
+        return result
+
+    def __apply_op(self, y, op):
```

Comment:
I guess we will need a strategy for ops with >2 arguments, and to correctly access their .share_ptr attribute?

Reply:
Hmm..for the moment I thought only for add, sub, mul. I think we can extend it after this PR?

Reply:
Yeah sounds good!